toplevel/opttoploop.ml

```diff
@@ -221,29 +277,65 @@ let execute_phrase print_outcome ppf phr =
       Compilenv.reset ~source_provenance:Timings.Toplevel
         ?packname:None !phrase_name;
       Typecore.reset_delayed_checks ();
+      let sstr, rewritten =
+        match sstr with
+        | [ { pstr_desc = Pstr_eval (e, attrs) ; pstr_loc = loc } ]
+        | [ { pstr_desc = Pstr_value (Asttypes.Nonrecursive,
+                                      [{ pvb_expr = e
+                                       ; pvb_pat = { ppat_desc = Ppat_any ; _ }
+                                       ; pvb_attributes = attrs
+                                       ; _ }])
+            ; pstr_loc = loc }
+          ] ->
+            let pat = Ast_helper.Pat.var (Location.mknoloc "_$") in
+            let vb = Ast_helper.Vb.mk ~loc ~attrs pat e in
+            [ Ast_helper.Str.value ~loc Asttypes.Nonrecursive [vb] ], true
+        | _ -> sstr, false
+      in
       let (str, sg, newenv) = Typemod.type_toplevel_phrase oldenv sstr in
       if !Clflags.dump_typedtree then Printtyped.implementation ppf str;
       let sg' = Typemod.simplify_signature sg in
+      (* Why is this done? *)
       ignore (Includemod.signatures oldenv sg sg');
       Typecore.force_delayed_checks ();
-      let res = Translmod.transl_store_phrases !phrase_name str in
+      let module_ident, res, size =
+        if Config.flambda then
+          let (module_ident, (res, size)) =
+            Translmod.transl_implementation_flambda !phrase_name
+              (str, Tcoerce_none)
+          in
+          remember module_ident 0 sg';
+          module_ident, close_phrase res, size
+        else
+          let size, res = Translmod.transl_store_phrases !phrase_name str in
+          Ident.create_persistent !phrase_name, res, size
+      in
       Warnings.check_fatal ();
       begin try
         toplevel_env := newenv;
-        let res = load_lambda ppf res in
+        let res = load_lambda ppf ~module_ident res size in
         let out_phr =
           match res with
           | Result v ->
-              Compilenv.record_global_approx_toplevel ();
```

Comment:
I don't think this currently work in non-flambda mode. I think that @trefis only fixed the flambda version.
But I will probably keep that part out of the 'enable-flambda' patch first.